paasta_tools/deployd/watchers.py

```diff
@@ -110,55 +108,145 @@ class SoaFileWatcher(PaastaWatcher):
-            default_proc_fun=YelpSoaEventHandler(filewatcher=self),
+        self.marathon_clients = get_marathon_clients_from_config()
+
+    def get_service_name_from_event(self, event):
+        """Get service_name from the file inotify event,
+        returns None if it is not an event we're interested in"""
+        starts_with = ['marathon-', 'deployments.json']
+        if any([event.name.startswith(x) for x in starts_with]):
+            service_name = event.path.split('/')[-1]
+        elif event.name.endswith('.json') and event.path.split('/')[-1] == 'secrets':
+            # this is needed because we put the secrets json files in a
+            # subdirectory so the service name would be "secrets" otherwise
+            service_name = event.path.split('/')[-2]
+        else:
+            service_name = None
+        return service_name
+
+    def process_event(self, event):
+        self.log.debug(event)
+        service_name = self.get_service_name_from_event(event)
+        if service_name:
+            self.log.info(f"Change of {event.name} in {event.path}")
+            self.bounce_service(service_name)
+
+    def bounce_service(self, service_name):
+        self.log.info(f"Checking if any instances for {service_name} need bouncing")
+        instances = list_all_instances_for_service(
+            service=service_name,
+            clusters=[self.filewatcher.cluster],
+            instance_type='marathon',
+            cache=False,
+        )
+        self.log.debug(instances)
+        service_instances = [(service_name, instance) for instance in instances]
+        service_instances = get_service_instances_needing_update(
+            self.marathon_clients,
+            service_instances,
+            self.filewatcher.cluster,
         )
+        for service, instance in service_instances:
+            self.log.info(f"{service}.{instance} has a new marathon app ID, and so needs bouncing")
+        service_instances_to_queue = [
+            # https://github.com/python/mypy/issues/2852
+            ServiceInstance(  # type: ignore
+                service=service,
+                instance=instance,
+                cluster=self.filewatcher.cluster,
+                bounce_by=int(time.time()),
+                watcher=type(self).__name__,
+                bounce_timers=None,
+                failures=0,
+            )
+            for service, instance in service_instances
+        ]
+        for service_instance in service_instances_to_queue:
+            self.filewatcher.inbox_q.put(service_instance)
 
     @property
     def mask(self):
+        mask = inotify.constants.IN_ALL_EVENTS
         boring_flags = ['IN_CLOSE_NOWRITE', 'IN_OPEN', 'IN_ACCESS', 'IN_ATTRIB']
-        return reduce(
-            lambda x, y: x | y,
-            [v for k, v in pyinotify.EventsCodes.OP_FLAGS.items() if k not in boring_flags],
-        )
+        for flag in boring_flags:
+            mask = mask - inotify.constants.boring_flag
```

Comment:
`mask = mask - getattr(inotify.constants, flag)`

Reply:
also probably cache the value or even pre-calculate it into a property during loading/initialization, doesn't seem like it changes ever

Reply:
+1, doesn't look quite right